File: src/main/java/org/lanternpowered/server/inventory/behavior/ShiftClickBehavior.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.inventory.behavior;

import org.lanternpowered.server.inventory.AbstractContainer;
import org.lanternpowered.server.inventory.AbstractInventorySlot;
import org.lanternpowered.server.inventory.AbstractSlot;
import org.lanternpowered.server.inventory.IInventory;

import org.checkerframework.checker.nullness.qual.Nullable;

@FunctionalInterface
public interface ShiftClickBehavior {

    /**
     * Gets the {@link IInventory}s that should be used when shift
     * clicking from the specified {@link AbstractInventorySlot}.
     * {@code null} can be returned to disable shift clicking.
     *
     * @param container The container
     * @param slot The slot
     * @return The target inventory, or null if none
     */
    @Nullable
    IInventory getTarget(AbstractContainer container, AbstractSlot slot);
}

File: src/main/kotlin/org/lanternpowered/api/item/inventory/behavior/ShiftClickBehavior.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.api.item.inventory.behavior

import org.lanternpowered.api.item.inventory.container.ExtendedContainer
import org.lanternpowered.api.item.inventory.ExtendedInventory
import org.lanternpowered.api.item.inventory.Inventory
import org.lanternpowered.api.item.inventory.slot.ExtendedSlot

/**
 * A behavior used when shift-clicking in a container.
 */
interface ShiftClickBehavior {

    /**
     * Gets the [ExtendedInventory] that should be used when shift clicking
     * on the specified [ExtendedSlot]. `null` can be returned to disable
     * shift clicking.
     *
     * @param container The container that is currently viewed
     * @param slot The slot that was shift-clicked on
     * @return The target inventory, or null if none
     */
    fun getTarget(container: ExtendedContainer, slot: ExtendedSlot): Inventory?
}
